typing/shape.ml

```diff
@@ -428,25 +473,69 @@ end) = struct
         Abs(x, read_back_force nf)
     | NStruct nstr ->
         Struct (Item.Map.map read_back_force nstr)
+    | NAlias nf -> Alias (read_back nf)
     | NProj (nf, item) ->
         Proj (read_back nf, item)
     | NLeaf -> Leaf
     | NComp_unit s -> Comp_unit s
     | NoFuelLeft t -> t
 
+  (* Sharing the memo tables is safe at the level of a compilation unit since
+    idents should be unique *)
+  let reduce_memo_table = Local_store.s_table Hashtbl.create 42
+  let read_back_memo_table = Local_store.s_table Hashtbl.create 42
+
   let reduce global_env t =
     let fuel = ref Params.fuel in
-    let reduce_memo_table = Hashtbl.create 42 in
-    let read_back_memo_table = Hashtbl.create 42 in
     let local_env = Ident.Map.empty in
     let env = {
       fuel;
       global_env;
-      reduce_memo_table;
-      read_back_memo_table;
+      reduce_memo_table = !reduce_memo_table;
+      read_back_memo_table = !read_back_memo_table;
       local_env;
     } in
     reduce_ env t |> read_back env
+
+  let weak_read_back env (nf : nf) : t =
+    let cache = Hashtbl.create 42 in
+    let rec weak_read_back env nf =
+      let memo_key = (env.local_env, nf) in
+      in_memo_table cache memo_key (weak_read_back_ env) nf
+    and weak_read_back_ env nf : t =
+      { uid = nf.uid;
+        desc = weak_read_back_desc env nf.desc;
+        approximated = nf.approximated }
+    and weak_read_back_desc env desc : desc =
+      let weak_read_back_no_force (Thunk { shape = t; _ }) = t in
```

Comment:
This feels dangerous, as it discards the `local_env`, which acts like a substitution on `t`, potentially leaving unbound variables in the output. What we would need here is to perform a parallel substitution on `t`, binding the variables that are bound by `local_env` to thre weak read back of the `nf` they are bound to.
Note that since `t` could contain sharing, this will need another hash table to ensure we do not break that sharing nor cause exponential slowdown.

Reply:
Right, that's a good catch. I am wondering whether this is an actual issue for the use that we have for this "weak" reduction: finding the `Uid` of a module without reducing all its components.

What exactly do you mean by "perform a parallel substitution" ? (we can maybe meet next week to discuss it ?)

Reply:
Worked-around in 93936096727b68a8dc42f83ec238f9d4b07327c6